keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -1118,6 +1119,49 @@ If a Pod references multiple claims managed by the same driver, then the driver
 can combine updating `podSchedulingContext.claims[*].unsuitableNodes` for all
 of them, after considering all claims.
 
+### Scheduled pods with unallocated or unreserved claims
+
+There are several scenarios where a Pod might be scheduled (= `pod.spec.nodeName`
+set) while the claims that it depends on are not allocated or not reserved for
+it:
+
+* A user might manually create a pod with `pod.spec.nodeName` already set.
+* Some special cluster might use its own scheduler and schedule pods without
+  using kube-scheduler.
+* The feature might have been disabled in kube-scheduler while scheduling
+  a pod with claims.
+
+The kubelet is refusing to run such pods and reports the situation through
+events and a pod condition (see below). It's an error scenario that should
+better be avoided.
+
+Users should avoid this situation by not scheduling pods manually. If they need
+it for some reason, they can use a node selector which matches only the desired
+node and then let kube-scheduler do the normal scheduling.
+
+Custom schedulers should emulate the behavior of kube-scheduler and ensure that
+claims are allocated and reserved before setting `pod.spec.nodeName`.
+
+The last scenario might occur during a downgrade or because of an
+administrator's mistake. Administrators can fix this by deleting such pods or
+ensuring that claims are usable by them. The latter is work that can be
+automated in kube-controller-manager:
+
+- If `pod.spec.nodeName` is set, kube-controller-manager can be sure that
+  kube-scheduler is not doing anything for the pod.
+- If such a pod has unallocated claims, kube-controller-manager can
+  create a `PodSchedulingContext` with only the `spec.selectedNode` field set
+  to the name of the node chosen for the pod. There is no need to list
+  suitable nodes because that choice is permanent, so resource drivers don't
+  need check for unsuitable nodes. All that they can do is to (re)try allocating
+  the claim until that succeeds.
+- If such a pod has allocated claims that are not reserved for it yet,
+  then kube-controller-manager can (re)try to reserve the claim until
+  that succeeds.
+
+Once all of those steps are complete, kubelet will notice that the claims are
+ready and run the pod.
```

Comment:
This probably belongs in a separate section or somewhere in the overall design - will check.

Reply:
Hrm, kubelet probanly reserves the right to fail any pod scheduled on it.  Eviction and graceful shutdown are general though and I agree those decisions don’t seem to need to specifically take into account DRA

Reply:
I read that as confirmation that the proposed change is useful because we should "try forever" and then having code that rectifies the situation helps.

Reply:
Correcting/clarifying my own words: I mean that components will continue trying to provision, run the pod, etc, instead of marking the Pod as Failed. Other mechanics (eviction, preemption, shutdown) would still Fail the pod, but that's beyond this KEP.

Reply:
Was there a conclusion about this?

Reply:
My understanding was that "try forever" is consistent with other kubelet error handling, in particular now that there even is a controller which can fix these error cases without having to delete the pod.

Is there any disagreement about that?

Reply:
I added:
> Until then it will keep checking periodically, just as
it does for other reasons that prevent a pod from running.

Reply:
LGTM